internal/db_impl/sqlite3/read_ops.go

```diff
@@ -0,0 +1,599 @@
+		if err != nil {
+			return nil, err
+		}
+
+		result = append(result, r...)
+	}
+
+	return result, nil
+}
+
+func (r readOps) MailboxFilterContains(ctx context.Context, mboxID imap.InternalMailboxID, messageIDs []db.MessageIDPair) ([]imap.InternalMessageID, error) {
+	return r.MailboxFilterContainsInternalID(ctx, mboxID, xslices.Map(messageIDs, func(t db.MessageIDPair) imap.InternalMessageID {
+		return t.InternalID
+	}))
+}
+
+func (r readOps) MailboxFilterContainsInternalID(ctx context.Context, mboxID imap.InternalMailboxID, messageIDs []imap.InternalMessageID) ([]imap.InternalMessageID, error) {
+	result := make([]imap.InternalMessageID, 0, len(messageIDs))
+
+	for _, chunk := range xslices.Chunk(messageIDs, db.ChunkLimit) {
+		query := fmt.Sprintf("SELECT `%v` FROM %v WHERE `%v` IN (%v) AND `%v` = ?",
+			v0.UIDsFieldMessageID,
+			v0.UIDsTableName,
+			v0.UIDsFieldMessageID,
+			GenSQLIn(len(chunk)),
+			v0.UIDsFieldMailboxID,
+		)
+
+		r, err := MapQueryRows[imap.InternalMessageID](ctx, r.qw, query, append(MapSliceToAny(chunk), mboxID)...)
+		if err != nil {
+			return nil, err
+		}
+
+		result = append(result, r...)
+	}
+
+	return result, nil
+}
+
+func (r readOps) GetMailboxCount(ctx context.Context) (int, error) {
+	query := fmt.Sprintf("SELECT COUNT(*) FROM %v", v0.MailboxesTableName)
+
+	return MapQueryRow[int](ctx, r.qw, query)
+}
+
+func (r readOps) GetMailboxMessageUIDsWithFlagsAfterAddOrUIDBump(ctx context.Context, mboxID imap.InternalMailboxID, messageIDs []imap.InternalMessageID) ([]db.UIDWithFlags, error) {
+	result := make([]db.UIDWithFlags, 0, len(messageIDs))
+
+	for _, chunk := range xslices.Chunk(messageIDs, db.ChunkLimit) {
+		query := fmt.Sprintf("SELECT `t1`.`remote_id`, GROUP_CONCAT(`t2`.`value`) AS `flags`, `ui_ds`.`recent`, `ui_ds`.`deleted`, `ui_ds`.`uid`, `ui_ds`.`uid_message` FROM `ui_ds`"+
+			" JOIN `messages` AS `t1` ON `ui_ds`.`uid_message` = `t1`.`id`"+
+			" LEFT JOIN `message_flags` AS `t2` ON `ui_ds`.`uid_message` = `t2`.`message_flags` WHERE `mailbox_ui_ds` = ? AND `uid_message` in (%v)"+
+			" GROUP BY `ui_ds`.`uid_message` ORDER BY `ui_ds`.`uid`",
+			GenSQLIn(len(chunk)))
+
+		args := make([]any, 0, len(chunk)+1)
+		args = append(args, mboxID)
+		args = append(args, MapSliceToAny(chunk)...)
+
+		r, err := MapQueryRowsFn(ctx, r.qw, query, func(scanner RowScanner) (db.UIDWithFlags, error) {
+			var r db.UIDWithFlags
+			var flags sql.NullString
+
+			if err := scanner.Scan(&r.RemoteID, &flags, &r.Recent, &r.Deleted, &r.UID, &r.InternalID); err != nil {
+				return db.UIDWithFlags{}, err
+			}
+
+			r.Flags = flags.String
+
+			return r, nil
+		}, args...)
+		if err != nil {
+			return nil, err
+		}
+
+		result = append(result, r...)
+	}
+
+	return result, nil
+}
+
+func (r readOps) MessageExists(ctx context.Context, id imap.InternalMessageID) (bool, error) {
+	query := fmt.Sprintf("SELECT COUNT(*) FROM %v WHERE `%v` = ?", v0.MessagesTableName, v0.MessagesFieldID)
```

Comment:
`SELECT COUNT(*)` is used in message exists and `SELECT id` is used in mailbox exists.
 
 Any reason why they should be same?
 
 Which one is more optimal?

Reply:
IIRC count on a specific column will not include null values, where count(*) will.

Reply:
Normalized exists queries with a better query.